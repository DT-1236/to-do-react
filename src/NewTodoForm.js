import React, { Component } from 'react';

class NewTodoForm extends Component {
  constructor(props) {
    super(props);
    this.state = { task: '' };
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }
  handleSubmit(evt) {
    evt.preventDefault();
    // do something with form data
    //this.props.addATask
    this.props.addATask(this.state);
    this.setState({ task: '' });
  }
  handleChange(evt) {
    // runs on every keystroke event
    this.setState({ task: evt.target.value });
  }
  render() {
    return (
      <form onSubmit={this.handleSubmit}>
        <label htmlFor="task">New Task:</label>
        <input
          name="task"
          value={this.state.task}
          onChange={this.handleChange}
        />
        <button>Add!</button>
      </form>
    );
  }
}
export default NewTodoForm;
