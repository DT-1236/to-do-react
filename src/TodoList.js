import React, { Component } from 'react';
import uuid from 'uuid/v4';
import NewTodoForm from './NewTodoForm';
import Todo from './Todo';

class TodoList extends Component {
  constructor(props) {
    super(props);
    //task is an object with uuid as the key
    this.state = { tasks: {} };
    this.creatNewTask = this.creatNewTask.bind(this);
    this.renderTasks = this.renderTasks.bind(this);
    this.removeTask = this.removeTask.bind(this);
    this.editTask = this.editTask.bind(this);
  }
  //create a new task
  creatNewTask({ task }) {
    // create a new obj in this.state.tasks
    // make a copy of the currst, add this obj to the curst, and setState witt the new state
    let curTasks = { ...this.state.tasks };
    curTasks = { ...curTasks, [uuid()]: task };
    this.setState({ tasks: curTasks });
  }

  // render a list of the list in the this.state
  renderTasks() {
    const result = Object.keys(this.state.tasks).map(id => (
      <Todo
        key={id}
        id={id}
        task={this.state.tasks[id]}
        remove={this.removeTask}
        editTask={this.editTask}
      />
    ));
    return result;
  }

  //delete function according to the id of the to do, which is also the key in this.state.tasks key
  removeTask(id) {
    let curSt = { ...this.state };
    let curTasks = curSt.tasks;
    delete curTasks[id];
    curSt = { tasks: curTasks };
    this.setState(curSt);
  }

  editTask(id, content) {
    //get the object
    console.log(id, content);
    let curTasks = { ...this.state.tasks };
    curTasks[id] = content;
    this.setState({ tasks: curTasks });
  }

  render() {
    return (
      <div>
        <NewTodoForm addATask={this.creatNewTask} />
        <ul>{this.renderTasks()}</ul>
      </div>
    );
  }
}
export default TodoList;
