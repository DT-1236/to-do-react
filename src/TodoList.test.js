import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import TodoList from './TodoList';
import NewTodoForm from './NewTodoForm';
//in src/**-test.js

//smokeTest

describe('simple smoking snapshots', function() {
  it('renders without crushing', () => {
    shallow(<TodoList />);
  });
  it('matches snapshot', function() {
    let wrapper = shallow(<TodoList />);
    let serialized = toJson(wrapper);
    expect(serialized).toMatchSnapshot();
  });
  it('initial state', () => {
    let wrapper = shallow(<TodoList />);
    expect(wrapper.state().tasks).toEqual({});
  });
});

describe('the todoList can do add, delete and ...', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(<TodoList />);
    let form = wrapper.find('form');
    let input = wrapper.find('input');
    input.simulate('change', { target: { name: 'task', value: 'test' } });
    form.simulate('submit');
  });
  afterEach(() => {
    wrapper.setState({
      tasks: {}
    });
  });

  it('the todoList can add a task', () => {
    expect(Object.keys(wrapper.state().tasks)).toHaveLength(1);
  });

  it('the todoList can remove a task', () => {
    let deleteBtn = wrapper.find('button[name="delete"]');
    deleteBtn.simulate('click');
    expect(Object.keys(wrapper.state().tasks)).toHaveLength(0);
  });

  it('the todoList can update a task', () => {
    let editButton = wrapper.find("button[name='edit']");
    editButton.simulate('click');
    let input = wrapper.find('input[name="content"]');
    input.simulate('change', { target: { name: 'content', value: 'updated' } });
    editButton.simulate('click');
    expect(Object.values(wrapper.state().tasks)[0]).toEqual('updated');
  });

  // it('edit mode, change the input', () => {
  //   let editButton = wrapper.find("button[name='edit']");
  //   editButton.simulate('click');
  //   expect(wrapper.state().isEdit).toBe(true);
  //   let input = wrapper.find('input');
  //   input.simulate('change', { target: { name: 'content', value: 'a' } });
  //   input = wrapper.find('input');
  //   expect(input.props().value).toBe('a');
  //   expect(wrapper.state().content).toBe('a');
  // });
});
