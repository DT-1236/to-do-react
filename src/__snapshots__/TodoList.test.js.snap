// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simple smoking snapshots matches snapshot 1`] = `
<div>
  <NewTodoForm
    addATask={[Function]}
  />
  <ul />
</div>
`;
